File: src/hmi/src/contract/topics.ts
/**
 * 10 个 MQTT 主题（SSOT 镜像：config/mqtt_topics.json）
 *
 * 主题模式：rgv/{devId}/{topic}（devId 由路径承载，不进入 payload）。
 * 方向语义（ADR-002）：cloud=云端/operator，device=控制器，hmi=安卓 App。
 * HMI 作为现场 operator 控制台，可发布 downlink（operator→device）与 hmi→cloud 主题。
 */
import mqttTopicsJson from '../../../../config/mqtt_topics.json';

export interface TopicMeta {
  topic: string;
  direction: string;
  qos: number;
  payload: string;
}

interface RawTopics {
  schema_version: string;
  source_of_truth: string;
  topic_pattern: string;
  note: string;
  topics: TopicMeta[];
}

export const MQTT_TOPICS_DOC = mqttTopicsJson as unknown as RawTopics;
export const TOPIC_PATTERN = MQTT_TOPICS_DOC.topic_pattern; // "rgv/{devId}/{topic}"
export const TOPIC_META_LIST: TopicMeta[] = MQTT_TOPICS_DOC.topics;
export const TOPIC_COUNT = TOPIC_META_LIST.length;

/** 10 个主题名（字面量联合） */
export type TopicName =
  | 'ota/cmd'
  | 'ota/data'
  | 'ota/progress'
  | 'ota/result'
  | 'param/set'
  | 'telemetry'
  | 'config/smdl'
  | 'interference/sync'
  | 'audit/log'
  | 'diag/log';

const META_BY_TOPIC: Record<string, TopicMeta> = {};
for (const t of TOPIC_META_LIST) META_BY_TOPIC[t.topic] = t;

export function getTopicMeta(name: string): TopicMeta | undefined {
  return META_BY_TOPIC[name];
}

/** 拼接完整主题路径：rgv/{devId}/{topic} */
export function topicFor(devId: string, name: TopicName | string): string {
  return `rgv/${devId}/${name}`;
}

/**
 * HMI 订阅（uplink / 双向中来自设备侧）：
 *  telemetry、ota/progress、ota/result、diag/log、interference/sync
 */
export const HMI_SUBSCRIBE: TopicName[] = [
  'telemetry',
  'ota/progress',
  'ota/result',
  'diag/log',
  'interference/sync',
];

/**
 * HMI 发布（downlink operator→device / hmi→cloud / 双向中来自 HMI 侧）：
 *  param/set、config/smdl、ota/cmd、ota/data、audit/log、interference/sync
 */
export const HMI_PUBLISH: TopicName[] = [
  'param/set',
  'config/smdl',
  'ota/cmd',
  'ota/data',
  'audit/log',
  'interference/sync',
];

/** OTA 触发相关主题（RBAC: admin） */
export const OTA_TOPICS: TopicName[] = ['ota/cmd', 'ota/data', 'ota/progress', 'ota/result'];
/** 参数/SMDL 配置相关主题（RBAC: engineer+） */
export const CONFIG_TOPICS: TopicName[] = ['param/set', 'config/smdl'];

File: src/ota/ota/topics.py
"""MQTT 主题构造与解析（对齐 config/mqtt_topics.json）。

契约要点
--------
- 主题路径 ``rgv/{devId}/{topic}``，**devId 由路径承载，不进入 payload**（ADR-002/治理）。
- 10 个主题见 mqtt_topics.json；OTA 使用其中 4 个：
  ota/cmd(cloud→device)、ota/data(cloud→device)、
  ota/progress(device→cloud)、ota/result(device→cloud)。
"""
from __future__ import annotations

from typing import List, Tuple

from . import config


def build_topic(dev_id: str, topic: str) -> str:
    """构造完整主题 ``rgv/{devId}/{topic}``。"""
    if not dev_id:
        raise ValueError("devId 不可为空")
    if not topic:
        raise ValueError("topic 不可为空")
    if "/" in dev_id:
        raise ValueError("devId 不得含 '/'（须由路径承载）")
    return config.TOPIC_PATTERN.format(devId=dev_id, topic=topic)


def parse_topic(full: str) -> Tuple[str, str]:
    """解析 ``rgv/{devId}/{topic}`` → (devId, topic)。失败抛 ValueError。"""
    parts = full.split("/")
    # rgv / devId / ...topic（topic 本身不含 '/')
    if len(parts) < 3 or parts[0] != "rgv":
        raise ValueError(f"非法主题路径: {full!r}")
    dev_id = parts[1]
    topic = "/".join(parts[2:])
    if not dev_id or not topic:
        raise ValueError(f"非法主题路径: {full!r}")
    return dev_id, topic


def ota_topics(dev_id: str) -> dict:
    """返回某设备的 4 个 OTA 完整主题。"""
    return {
        "cmd": build_topic(dev_id, config.TOPIC_OTA_CMD),
        "data": build_topic(dev_id, config.TOPIC_OTA_DATA),
        "progress": build_topic(dev_id, config.TOPIC_OTA_PROGRESS),
        "result": build_topic(dev_id, config.TOPIC_OTA_RESULT),
    }


def is_ota_topic(topic: str) -> bool:
    try:
        _, rel = parse_topic(topic)
    except ValueError:
        return False
    return rel in {
        config.TOPIC_OTA_CMD,
        config.TOPIC_OTA_DATA,
        config.TOPIC_OTA_PROGRESS,
        config.TOPIC_OTA_RESULT,
    }


def all_topic_relpaths() -> List[str]:
    """从 SSOT 配置读出全部 10 个相对主题（校验：数量=10）。"""
    cfg = config.load_topics_config()
    topics = [t["topic"] for t in cfg["topics"]]
    assert len(topics) == config_load_topic_count(), "主题数量与 SSOT 不符"
    return topics


def config_load_topic_count() -> int:
    return 10
